### ISSUE TITLE
The update equation in examples/generative /real_nvp.py

### ISSUE THREAD

**Issue by amida47** [keras-team/keras-io#1339]
in your implementation of a realNVP, here is a part of the model class
```
class RealNVP(keras.Model):
    def __init__(self, num_coupling_layers):
        super().__init__()

        self.num_coupling_layers = num_coupling_layers

        # Distribution of the latent space.
        self.distribution = tfp.distributions.MultivariateNormalDiag(
            loc=[0.0, 0.0], scale_diag=[1.0, 1.0]
        )
        self.masks = np.array(
            [[0, 1], [1, 0]] * (num_coupling_layers // 2), dtype="float32"
        )
        self.loss_tracker = keras.metrics.Mean(name="loss")
        self.layers_list = [Coupling(2) for i in range(num_coupling_layers)]

    @property
    def metrics(self):
        """List of the model's metrics.

        We make sure the loss tracker is listed as part of `model.metrics`
        so that `fit()` and `evaluate()` are able to `reset()` the loss tracker
        at the start of each epoch and at the start of an `evaluate()` call.
        """
        return [self.loss_tracker]

    def call(self, x, training=True):
        log_det_inv = 0
        direction = 1
        if training:
            direction = -1
        for i in range(self.num_coupling_layers)[::direction]:
            x_masked = x * self.masks[i]
            reversed_mask = 1 - self.masks[i]
            s, t = self.layers_list[i](x_masked)
            s *= reversed_mask
            t *= reversed_mask
            gate = (direction - 1) / 2
            x = (
                reversed_mask
                * (x * tf.exp(direction * s) + direction * t * tf.exp(gate * s))
                + x_masked
            )
            log_det_inv += gate * tf.reduce_sum(s, [1])

        return x, log_det_inv

    # Log likelihood of the normal distribution plus the log determinant of the jacobian.
```
my problem is with the values the variable direction takes, if training is true direction = -1, but with this value if we calculate the new value of x (in this case lets take x=[x1, x2] a two dimension data vector) going forward (from data to latent space), I found that given the current mask is [1, 0] and that s = [s1,s2] and t = [t1, t2] and :
```
x = (
                reversed_mask
                * (x * tf.exp(direction * s) + direction * t * tf.exp(gate * s))
                + x_masked
            )

```
will give me x = [x1 , x2 * exp(-s2) - t2 *exp(-s2)] which different than what the formula says x should equal to [x1, x2 * exp(s2) + t2]

if you can please clarify the choice if setting direction values

**Comment by amida47** [comment]
since @mbrukman was last to modify this file, I hope you can clarify my confusion

**Comment by maitry63** [comment]
Hi @amida47 , 
The confusion comes from the fact that, in normalizing flows such as RealNVP, the terms **forward** and **inverse** refer to the direction of the **invertible transformation**, rather than to the usual notion of a neural network forward pass.

During maximum-likelihood training, we start from a data sample (x) and compute its latent representation (z) by applying the **inverse transformation** (f^{-1}(x)). This is required because the data likelihood is computed using the change-of-variables formula,

`[log (pX(x)) = log(pZ f(x)) + log(|det(∂f(x)/∂xT))]`

where (z=f^{-1}(x)).

Conversely, during generation (sampling), we draw (z) from the latent distribution and apply the **forward transformation** (f(z)) to generate a data sample (x).

For a coupling layer with mask `[1, 0]`, the two transformations are:

* **Forward (latent → data):**
  [x_2 = z_2,\exp(s_2) + t_2]

* **Inverse (data → latent):**
  [z_2 = (x_2 - t_2)\exp(-s_2) = x_2\exp(-s_2) - t_2\exp(-s_2)]

Your derived expression,

[x_2\exp(-s_2) - t_2\exp(-s_2),]

therefore corresponds exactly to the inverse transformation, which is the one used during training.

In this implementation, the `call()` method is written so that it can perform either  transformation. When `training=True`, it sets `direction = -1` and iterates through the coupling layers in reverse order, implementing the inverse mapping (data → latent). When `training=False` (for example, when `model.predict()` is called), it sets `direction = 1` and applies the forward mapping (latent → data). This allows the same implementation to be reused for both likelihood computation and sample generation.

This behavior is also reflected later in the  [tutorial](https://keras.io/examples/generative/real_nvp/) -

```python
# From data to latent space.
z, _ = model(normalized_data)

# From latent space to data.
samples = model.distribution.sample(3000)
x, _ = model.predict(samples)
```

where `model(normalized_data)` computes the inverse transformation, while `model.predict(samples)` computes the forward transformation.

For additional details on the forward and inverse coupling transformations and the change-of-variables formulation, see the original RealNVP paper: https://arxiv.org/abs/1605.08803.
